### PR TITLE
Add lshift and rshift operators

### LINKED ISSUE #7727: Implement lshift and rshift operators
### Is your feature request related to a problem?

I've noticed that xarray does not support the lshift or rshift operators.  While this is workable by using the numpy array the addition of the operators also seem somewhat trivial and reduces friction for numpy users adopting xarray.   

Is the exclusion of these operators intentional or oversight or some other historic reason?

### Describe the solution you'd like

Implement the operators if it makes sense. 

### Describe alternatives you've considered

_No response_

### Additional context

I experimented with the potential edits here: https://github.com/abrammer/xarray/tree/feat/add_lshift_and_rshift_operators
but don't want to open a PR without an initial discussion.   I haven't delved into this area of xarray before so I probably missed something or did something wrong.

### PR DESCRIPTION
**Pull request by abrammer**
This seems a little too easy, but the PR just adds lshift and rshit to `generate_ops.py` and `ops.py` then generated the 2 stub files with `generate_ops.py`. 

- [x] Closes #7727
- [x] Tests added for:
   -  xr.Variable
   -  DataArray
   - dask
   - groupby
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
   - There will be new `<<` `>>` syntax for dataarrays and datasets. 
- [x] New functions/methods are listed in `api.rst`
    - Existing basic math operators aren't really listed, that I can find. 
    - Could add example code in the  "Data arrays also implement many numpy.ndarray methods:" section of computation.rst ?

I'll wait for initial review before spending time on whats-new and api docs. 

This enables bit shifting on arrays (when the type supports it).  

```python
import xarray as xr

x = xr.DataArray(5)
print(x << 1)
# <xarray.DataArray ()>
# array(10)


x = xr.DataArray([0,1,2,3,4,5])
print(x << 1)
# <xarray.DataArray (dim_0: 6)>
# array([ 0,  2,  4,  6,  8, 10])
# Dimensions without coordinates: dim_0


x = xr.DataArray([0,1,2,3,4,5])
print(x >> 1)
# <xarray.DataArray (dim_0: 6)>
# array([0, 0, 1, 1, 2, 2])
# Dimensions without coordinates: dim_0


x = xr.DataArray([0,1,2,3,4,5.])
print(x >> 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspaces/xarray/xarray/core/_typed_ops.py", line 255, in __rshift__
    return self._binary_op(other, operator.rshift)
  File "/workspaces/xarray/xarray/core/dataarray.py", line 4566, in _binary_op
    f(self.variable, other_variable)
  File "/workspaces/xarray/xarray/core/_typed_ops.py", line 461, in __rshift__
    return self._binary_op(other, operator.rshift)
  File "/workspaces/xarray/xarray/core/variable.py", line 2662, in _binary_op
    f(self_data, other_data) if not reflexive else f(other_data, self_data)
TypeError: ufunc 'right_shift' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### CODE REVIEW

**Comment by dcherian** [comment]
> I don't see tests for other ops. Are these tested somewhere if so I can add tests when I find them.

Grepping for `unary_op` and `binary_op` shows a bunch in `test_dataset.py`, `test_dataarray.py`, `test_variable.py`

**Comment by dcherian** [comment]
I forgot to say, this looks pretty great already. We just need tests.

Thank you!

File: xarray/tests/test_dask.py [hunk 1]
```diff
@@ -178,6 +178,20 @@ def test_binary_op(self):
         self.assertLazyAndIdentical(u + u, v + v)
         self.assertLazyAndIdentical(u[0] + u, v[0] + v)
 
+    def test_binary_op_bitshift(self):
```
Comment (Illviljan) [inline xarray/tests/test_dask.py:181]:
```suggestion
    def test_binary_op_bitshift(self) -> None:
```
Add None so mypy also checks the test.

File: xarray/tests/test_variable.py [hunk 2]
```diff
@@ -344,9 +344,10 @@ def test_pandas_period_index(self):
         assert v[0] == pd.Period("2000", freq="B")
         assert "Period('2000-01-03', 'B')" in repr(v)
 
-    def test_1d_math(self):
-        x = 1.0 * np.arange(5)
-        y = np.ones(5)
+    @pytest.mark.parametrize("dtype", [float, int])
+    def test_1d_math(self, dtype):
```
Comment (Illviljan) [inline xarray/tests/test_variable.py:348]:
```suggestion
    def test_1d_math(self, dtype: np.typing.DTypeLike) -> None:
```

Comment (abrammer) [reply xarray/tests/test_variable.py:348]: Adding type checking on `test_1d_math` triggered attribute errors on existing lines of the test.  
`xarray/tests/test_variable.py:370: error: "ndarray[Any, dtype[bool_]]" has no attribute "values" `

I'm not sure of the simple fix so removed it from that test.

Comment (Illviljan) [reply xarray/tests/test_variable.py:348]: Not sure why numpy takes priority when it's first in the operation. 
If you don't want to work on it, add the typing back but add ignores to the bad lines:

```python
xv.values  # type: ignore[attr-defined] # TODO: Fix mypy thinking numpy takes priority, GH7780
```

File: xarray/tests/test_groupby.py [hunk 3]
```diff
@@ -808,6 +808,29 @@ def test_groupby_math_more() -> None:
         ds + ds.groupby("time.month")
 
 
+def test_groupby_math_bitshift() -> None:
+    ds = Dataset(
+        {
+            "x": ("level", np.ones(4, dtype=int)),
+            "y": ("level", np.ones(4, dtype=int) * -1),
+            "level": [0, 1, 2, 3],
```
Comment (dcherian) [inline xarray/tests/test_groupby.py:816]:
```suggestion
            "level": [0, 1, 2, 2],
```

So that we actually test the groupby  bit.

Comment (abrammer) [reply xarray/tests/test_groupby.py]: ah yeah, good point.  Had to shuffle it a little bit more but think that covers the actual groupby operation better now.

**Comment by abrammer** [comment]
The commits yesterday were to add an entry to whats-new and a couple examples lines to the computation doc page.  I didn't find the binary_ops listed in methods anywhere, so this was the best idea I had? In the block just above missing-values: https://xray--7741.org.readthedocs.build/en/7741/user-guide/computation.html#missing-values

Otherwise, I think this is done from my perspective.

**Comment by dcherian** [comment]
Thanks @abrammer very nice work!

**Pull request merged**